File: src/main/java/com/github/patrick/utils/bukkit/BukkitClassLoader.java
package com.github.patrick.utils.bukkit;

import org.apache.commons.lang3.ClassUtils;
import org.apache.commons.lang3.StringUtils;
import org.apache.commons.lang3.reflect.ConstructorUtils;
import org.jetbrains.annotations.NotNull;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.List;

@SuppressWarnings({"rawtypes", "unused"})
public final class BukkitClassLoader {
    private BukkitClassLoader() {}

    @SuppressWarnings("unchecked")
    public static <T> T load(@NotNull final Class<T> type, @NotNull final String prefix, @NotNull final String version, @NotNull final Object... initialArgs) throws UnsupportedOperationException {
        String packageName = type.getPackage().getName();
        String className = prefix + type.getSimpleName();
        int lastDot = packageName.lastIndexOf('.');
        Class[] parameterTypes = ClassUtils.toClass(initialArgs);
        List<String> candidates = new ArrayList<>(2);
        candidates.add(packageName + '.' + version + '.' + className);

        if (lastDot > 0) {
            String superPackageName = packageName.substring(0, lastDot);
            String subPackageName = packageName.substring(lastDot + 1);
            candidates.add(superPackageName + '.' + version + '.' + subPackageName + '.' + className);
        }

        try {
            Class<? extends T> nmsClass = null;
            for (String name : candidates) {
                try { nmsClass = Class.forName(name, true, type.getClassLoader()).asSubclass(type); } catch (ClassNotFoundException ignored) {}
            }
            if (nmsClass == null) throw new ClassNotFoundException("Not found nms library class: " + candidates.toString());
            Constructor<?> constructor = ConstructorUtils.getMatchingAccessibleConstructor(nmsClass, parameterTypes);

            if (constructor == null)
                throw new UnsupportedOperationException(type.getName() + " does not have Constructor for [" + StringUtils.join(parameterTypes, ", ") + "]");
            return (T) constructor.newInstance(initialArgs);
        } catch (ClassNotFoundException e) {
            throw new UnsupportedOperationException(type.getName() + " does not support this version " + version, e);
        } catch (IllegalAccessException e) {
            throw new UnsupportedOperationException(type.getName() + " constructor is not visible");
        } catch (InstantiationException e) {
            throw new UnsupportedOperationException(type.getName() + " is abstract class");
        } catch (InvocationTargetException e) {
            throw new UnsupportedOperationException(type.getName() + " has an error occurred while creating the instance", e);
        }
    }

    @NotNull
    public static String getBukkitVersion(@NotNull Object bukkitClass) {
        String packageName = bukkitClass.getClass().getPackage().getName();
        return packageName.substring(packageName.lastIndexOf(".") + 1);
    }
}

File: src/main/kotlin/com/github/patrick/utils/bukkit/BukkitClassLoader.kt
package com.github.patrick.utils.bukkit

import org.apache.commons.lang3.ClassUtils.toClass
import org.apache.commons.lang3.reflect.ConstructorUtils.getMatchingAccessibleConstructor

import java.util.ArrayList

@Suppress("unused")
object BukkitClassLoader {
    @Throws(ClassNotFoundException::class, UnsupportedOperationException::class, UnsupportedClassVersionError::class, IllegalAccessError::class, InstantiationError::class)
    fun <T> load(classType: Class<T>, prefix: String, version: String, vararg initialArgs: Any?): T {
        val packageName = classType.`package`.name
        val className = "$prefix${classType.simpleName}"
        val candidates: MutableList<String> = ArrayList(2)
        val lastDot = packageName.lastIndexOf('.')
        var nmsClass: Class<out T>? = null
        candidates.add("$packageName.$version.$className")
        if (lastDot > 0) candidates.add("${packageName.substring(0, lastDot)}.$version.${packageName.substring(lastDot + 1)}.$className")
        return try {
            candidates.forEach {
                try {
                    nmsClass = nmsClass?: Class.forName(it, true, classType.classLoader).asSubclass(classType)
                } catch (ignored: ClassNotFoundException) {}
            }
            (getMatchingAccessibleConstructor(
                (nmsClass?: throw ClassNotFoundException("Not found class: $candidates")),
                *toClass(initialArgs))?: throw UnsupportedOperationException("${classType.name} does not have constructor for [${toClass(initialArgs).joinToString()}]"
            )).newInstance(initialArgs)
        } catch (e: ClassNotFoundException) { throw UnsupportedClassVersionError("${classType.name} does not support this version '$version'")
        } catch (e: IllegalAccessException) { throw IllegalAccessError("${classType.name} constructor is not visible")
        } catch (e: InstantiationException) { throw InstantiationError("${classType.name} is abstract class or an interface") }
    }
}